File: backend/src/models/thread.model.ts
import mongoose from 'mongoose';

const reactSchema = new mongoose.Schema({
    reactIcon : {type: String, required: true}, // The path to the icon of the reaction
    username : {type: String, required: true},
    timestamp: { type: Date, default: Date.now }
});

const messageSchema = new mongoose.Schema({
    username: { type: String, required: true }, // The username of the user who sent the message
    timestamp: { type: Date, default: Date.now },
    content: { type: String, required: true },
    reactions : {type: [reactSchema], default: []}, // Array of reactions
    respondsTo : {type: String, default: ''}, // The id of the message this message is responding to
});

const threadSchema = new mongoose.Schema({
    bookTitle: { type: String, required: true },
    image: { type: String },
    username: { type: String, required: true },
    title: { type: String, required: true },
    timestamp: { type: Date, default: Date.now },
    messages: { type: [messageSchema], default: [] }
});

// Create a text index on bookTitle, title, and username
threadSchema.index({ bookTitle: 'text', title: 'text', username: 'text' });

const Thread = mongoose.model('Thread', threadSchema, "threads");

export default Thread;

File: app/lib/models/thread_model.dart
class React {
  final String id;
  final String reactIcon;
  final String username;
  final DateTime timestamp;

  React({
    required this.id,
    required this.reactIcon, 
    required this.username,
    required this.timestamp,
  });

  factory React.fromJson(Map<String, dynamic> json) {
    return React(
      id: json['_id'],
      reactIcon: json['reactIcon'],
      username: json['username'],
      timestamp: DateTime.parse(json['timestamp']),
    );
  }
}

class Message {
  final String id;
  final String username;
  final DateTime timestamp;
  final String content;
  final List<React> reactions;
  final String? respondsTo;

  Message({
    required this.id,
    required this.username,
    required this.timestamp,
    required this.content,
    required this.reactions,
    this.respondsTo,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    var reactionsList = json['reactions'] as List;
    List<React> reactions = reactionsList.map((i) => React.fromJson(i)).toList();

    return Message(
      id: json['_id'],
      username: json['username'],
      timestamp: DateTime.parse(json['timestamp']),
      content: json['content'],
      reactions: reactions,
      respondsTo: json['respondsTo'],
    );
  }
}

class Thread {
  final String id;
  final String bookTitle;
  final String username;
  final String title;
  final String? image;
  final DateTime timestamp;
  final List<Message> messages;

  Thread({
    required this.id,
    required this.bookTitle,
    required this.username,
    required this.title,
    required this.timestamp,
    required this.messages,
    this.image,
  });

  factory Thread.fromJson(Map<String, dynamic> json) {
    var messagesList = json['messages'] as List;
    List<Message> messages = messagesList.map((i) => Message.fromJson(i)).toList();

    return Thread(
      id: json['_id'],
      bookTitle: json['bookTitle'],
      username: json['username'],
      title: json['title'],
      timestamp: DateTime.parse(json['timestamp']),
      messages: messages,
      image: json['image'],
    );
  }
}